src/components/tl/cuda/alltoall/alltoall_ce.c

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "alltoall.h"
+#include "core/ucc_mc.h"
+#include "tl_cuda_cache.h"
+
+ucc_status_t ucc_tl_cuda_alltoall_ce_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_mc_ee_destroy_event((void*)task->alltoall_ce.copy_done,
+                            UCC_EE_CUDA_STREAM);
+    ucc_tl_cuda_task_put(task);
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_cuda_alltoall_ce_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task   = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team   = TASK_TEAM(task);
+    ucc_tl_cuda_sync_t *sync   = TASK_SYNC(task, team->rank);
+    ucc_rank_t          n_done = 0;
+    ucc_tl_cuda_sync_t *peer_sync;
+    ucc_status_t status;
+    ucc_rank_t i;
+
+    status = ucc_mc_ee_event_test(task->alltoall_ce.copy_done,
+                                  UCC_EE_CUDA_STREAM);
+    if (status != UCC_OK) {
+        task->super.super.status = status;
+        goto exit;
+    }
+    sync->seq_num[1] = task->seq_num;
+    __sync_synchronize();
+    asm volatile("": : :"memory");
+    for (i = 0; i < team->size; i++) {
+        peer_sync = TASK_SYNC(task, i);
+        if (peer_sync->seq_num[1] == task->seq_num) {
+            n_done++;
+        }
+    }
+    task->super.super.status = UCC_INPROGRESS;
+    if (n_done == team->size) {
+        task->super.super.status = UCC_OK;
+    }
+exit:
+    return task->super.super.status;
+}
+
+ucc_status_t ucc_tl_cuda_alltoall_setup(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_coll_args_t    *args = &TASK_ARGS(task);
+    ucc_tl_cuda_sync_t *sync = TASK_SYNC(task, team->rank);
+    volatile ucc_tl_cuda_sync_t *peer_sync;
+    ucc_tl_cuda_cache_t *cache;
+    ucc_mem_attr_t mem_attr;
+    ucc_status_t status;
+    size_t data_len;
+    ucc_rank_t i;
+
+    data_len = ucc_dt_size(args->src.info.datatype) * args->src.info.count;
```

Comment:
maybe all these lines 67-81 can go to some helper fn ucc_tl_cuda_map_local_buffer.
I recall the same code will be used for alltoallv

Reply:
yes, it makes sense, but I'm not sure how the interface for such ucc_tl_cuda_map_local_buffer should look like. Let's do it for next PR when we will have another collective?